### PR TITLE
[SPARK-28330][SQL] Support ANSI SQL: result offset clause in query expression

### PR DESCRIPTION
**Pull request by beliefer**
## What changes were proposed in this pull request?

This is a ANSI SQL and feature id is `F861`
```
<query expression> ::=
[ <with clause> ] <query expression body>
[ <order by clause> ] [ <result offset clause> ] [ <fetch first clause> ]

<result offset clause> ::=
OFFSET <offset row count> { ROW | ROWS }
```
For example:
```
SELECT customer_name, customer_gender FROM customer_dimension 
   WHERE occupation='Dancer' AND customer_city = 'San Francisco' ORDER BY customer_name;
    customer_name     | customer_gender
----------------------+-----------------
 Amy X. Lang          | Female
 Anna H. Li           | Female
 Brian O. Weaver      | Male
 Craig O. Pavlov      | Male
 Doug Z. Goldberg     | Male
 Harold S. Jones      | Male
 Jack E. Perkins      | Male
 Joseph W. Overstreet | Male
 Kevin . Campbell     | Male
 Raja Y. Wilson       | Male
 Samantha O. Brown    | Female
 Steve H. Gauthier    | Male
 William . Nielson    | Male
 William Z. Roy       | Male
(14 rows)

SELECT customer_name, customer_gender FROM customer_dimension 
   WHERE occupation='Dancer' AND customer_city = 'San Francisco' ORDER BY customer_name OFFSET 8;
   customer_name   | customer_gender
-------------------+-----------------
 Kevin . Campbell  | Male
 Raja Y. Wilson    | Male
 Samantha O. Brown | Female
 Steve H. Gauthier | Male
 William . Nielson | Male
 William Z. Roy    | Male
(6 rows)
```
There are some mainstream database support the syntax.
**PostgreSQL:**
https://www.postgresql.org/docs/11/queries-limit.html

**Vertica:**
https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Statements/SELECT/OFFSETClause.htm?zoom_highlight=offset

**MySQL:**
https://dev.mysql.com/doc/refman/5.6/en/select.html

## How was this patch tested?
new UT.
There are some show of the PR on my production environment.
```
spark-sql> select * from gja_test_partition;
a       A       ao      1
b       B       bo      1
c       C       co      1
d       D       do      1
e       E       eo      2
g       G       go      2
h       H       ho      2
j       J       jo      2
f       F       fo      3
k       K       ko      3
l       L       lo      4
i       I       io      4
Time taken: 6.618 s
spark-sql> select * from gja_test_partition offset 3;
d       D       do      1
e       E       eo      2
g       G       go      2
h       H       ho      2
j       J       jo      2
f       F       fo      3
k       K       ko      3
l       L       lo      4
i       I       io      4
Time taken: 6.368 s
spark-sql> select * from gja_test_partition limit 5 offset 3;
d       D       do      1
e       E       eo      2
g       G       go      2
h       H       ho      2
j       J       jo      2
Time taken: 25.141 s
spark-sql> select * from gja_test_partition order by key;
a       A       ao      1
b       B       bo      1
c       C       co      1
d       D       do      1
e       E       eo      2
f       F       fo      3
g       G       go      2
h       H       ho      2
i       I       io      4
j       J       jo      2
k       K       ko      3
l       L       lo      4
Time taken: 16.894 s
spark-sql> select * from gja_test_partition order by key offset 3;
d       D       do      1
e       E       eo      2
f       F       fo      3
g       G       go      2
h       H       ho      2
i       I       io      4
j       J       jo      2
k       K       ko      3
l       L       lo      4
Time taken: 19.191 s
spark-sql> select * from gja_test_partition order by key limit 5 offset 3;
d       D       do      1
e       E       eo      2
f       F       fo      3
g       G       go      2
h       H       ho      2
Time taken: 12.556 s
```

### CODE REVIEW

**Comment by beliefer** [comment]
Retest this please.

**Comment by beliefer** [comment]
@dongjoon-hyun Could you take a look at this PR?

**Comment by beliefer** [comment]
cc: @gatorsmile @dongjoon-hyun

**Comment by beliefer** [comment]
@dongjoon-hyun @HyukjinKwon Could you help me review this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ */
+case class OffsetExec(offset: Int, child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputPartitioning: Partitioning = SinglePartition
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+
+  override def executeCollect(): Array[InternalRow] = child.executeCollect.drop(offset)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: Can we avoid collect everything into driver side?

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: Do you have a good suggestion?

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: @HyukjinKwon I have resolved this issue collect results to driver.

**Comment by beliefer** [comment]
Retest this please.

**Comment by beliefer** [comment]
Retest this please.

**Comment by beliefer** [comment]
Retest this please.

**Comment by beliefer** [comment]
@dongjoon-hyun @HyukjinKwon Could you help me to review this PR?

**Comment by gatorsmile** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 2]
```diff
@@ -1520,6 +1523,7 @@ NOT: 'NOT' | '!';
 NULL: 'NULL';
 NULLS: 'NULLS';
 OF: 'OF';
+OFFSET: 'OFFSET';
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: We should also update doc/sql-keywords.md

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 3]
```diff
@@ -399,6 +399,7 @@ queryOrganization
       (SORT BY sort+=sortItem (',' sort+=sortItem)*)?
       windowClause?
       (LIMIT (ALL | limit=expression))?
+      (OFFSET offset=expression)?
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: does SQL standard define the order of LIMIT and OFFSET clauses? Here we force users to write LIMIT clause first.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+case class OffsetExec(offset: Int, child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = child.output
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    val rdd = child.execute()
+    val arr = rdd.take(offset)
+    rdd.filter(!arr.contains(_))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: We need some design about how to implement OFFSET on a distributed data set. The current approach works, but it's not robust as it needs to broadcast `arr` (which can be large if OFFSET is large) and relies on object equality.

We may need to follow how LIMIT is implemented in Spark.

**Comment by cloud-fan** [comment]
Do we have to support all the features standard SQL supports? This one is not easy to implement and I don't know how many users want it. Does other distributed SQL engine support it? e.g. Hive, Presto, Impala, BigQuery, etc.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Yes, SQL standard defined the order for `LIMIT` and `OFFSET` clauses. `LIMIT` at the front of `OFFSET`.
The behavior looks contrary to MySQL direction. 
`MySQL` support `LIMIT 10, OFFSET 10`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala (hunk 4 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: Yes, I think so.
I have referenced the implement of `LIMIT`, but `OFFSET` looks can't follow the same way as `LIMIT`.
Second, `OFFSET` easier to generate large amounts of data than `LIMIT`.
I have an immature suggestion give a limitation on 'OFFSET '.

**Comment by beliefer** [comment]
Impala and BigQuery support `OFFSET`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: A rough idea:
1. get the numRecords of the first partition
2. If the numRecords is bigger than OFFSET, go to step 4
3. get numRecords of more partitions (quadruple and retry like LIMIT), until total numRecords is bigger than OFFSET.
4. Now we have the numRecords of some head partitions that totoal numRecords exceeds the OFFSET, we can easily skip the head records.

If we have accurate per-partition numRecords statistics, we can go step 4 directly.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: @cloud-fan It is a better idea for me. I will reactor the current implement.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: @cloud-fan There are exists a problem the index of partition and the order of data are inconsistent.
I have a new implement but not works file as I can't assurance the order of output produced by child plans.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/offset.scala]: So, I want know how to assurance the order?

**Comment by findepi** [comment]
Presto supports `OFFSET`.

**Comment by beliefer** [comment]
> Presto supports `OFFSET`.

Thanks for your investigate

**Comment by cloud-fan** [comment]
@beliefer are you still working on it?

**Comment by beliefer** [comment]
> @beliefer are you still working on it?

Yes
https://github.com/beliefer/spark/commit/0b52681150cb047e54d97677730203c98b4fe86d

**Comment by beliefer** [comment]
@gatorsmile @cloud-fan @HyukjinKwon @dongjoon-hyun @findepi 
Because the PR too old, there exists massive conflict which is hard to resolve. So I open this new PR to support this feature.
The new PR is https://github.com/apache/spark/pull/27237

**Pull request closed without merging**